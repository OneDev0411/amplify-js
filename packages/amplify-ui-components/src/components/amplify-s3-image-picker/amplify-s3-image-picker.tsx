import { Component, Prop, h, State, Host } from '@stencil/core';
import { NO_STORAGE_MODULE_FOUND } from '../../common/constants';
import { AccessLevel } from '../../common/types/storage-types';
import { Storage } from '@aws-amplify/storage';
import { Logger } from '@aws-amplify/core';
import { calcKey, getStorageObject } from '../../common/helpers';

const logger = new Logger('S3ImagePicker');

@Component({
  tag: 'amplify-s3-image-picker',
})
export class AmplifyS3ImagePicker {
  /* String representing directory location to image file */
  @Prop() path: string;
  /* The content type header used when uploading to S3 */
  @Prop() contentType: string = 'binary/octet-stream';
  /* The access level of the image */
  @Prop() level: AccessLevel = AccessLevel.Public;
  /* Whether or not to use track the get/put of the image */
  @Prop() track: boolean;
  /* Cognito identity id of the another user's image */
  @Prop() identityId: string;
  /* Callback used to generate custom key value */
  @Prop() fileToKey: (data: object) => string;
  @Prop() headerTitle?: string = 'Add Profile Photos';
  @Prop() headerHint?: string = 'Ancilliary text or content may occupy this space here';
  @Prop() placeholderHint?: string = 'Placeholder hint';
  @Prop() buttonText?: string = 'Upload';

  @State() src: string | object;

  async handlePick(file) {
    const { path = '', level, track, identityId, fileToKey } = this;
    const key = path + calcKey(file, fileToKey);

    if (!Storage || typeof Storage.put !== 'function') {
      throw new Error(NO_STORAGE_MODULE_FOUND);
    }

    try {
      const data = await Storage.put(key, file, {
        level,
        contentType: file['type'],
        track,
      });
      logger.debug(data);
      this.src = await getStorageObject(key, level, track, identityId, logger);
    } catch (error) {
      logger.error(error);
      throw new Error(error);
    }
  }

  render() {
    return (
      <Host>
        <slot name="photo-picker">
          <amplify-photo-picker
            previewSrc={this.src}
            onClickHandler={this.handlePick}
            headerTitle={this.headerTitle}
            headerHint={this.headerHint}
            placeholderHint={this.placeholderHint}
            buttonText={this.buttonText}
          ></amplify-photo-picker>
        </slot>
      </Host>
    );
  }
}
