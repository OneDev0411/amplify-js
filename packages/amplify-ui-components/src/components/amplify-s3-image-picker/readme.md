# amplify-s3-image-picker



<!-- Auto Generated Below -->


## Properties

| Property          | Attribute          | Description | Type                                                                 | Default                                                |
| ----------------- | ------------------ | ----------- | -------------------------------------------------------------------- | ------------------------------------------------------ |
| `buttonText`      | `button-text`      |             | `string`                                                             | `I18n.get(Translations.IMAGE_PICKER_BUTTON_TEXT)`      |
| `contentType`     | `content-type`     |             | `string`                                                             | `'binary/octet-stream'`                                |
| `fileToKey`       | --                 |             | `(data: object) => string`                                           | `undefined`                                            |
| `headerHint`      | `header-hint`      |             | `string`                                                             | `I18n.get(Translations.IMAGE_PICKER_HINT)`             |
| `headerTitle`     | `header-title`     |             | `string`                                                             | `I18n.get(Translations.IMAGE_PICKER_TITLE)`            |
| `identityId`      | `identity-id`      |             | `string`                                                             | `undefined`                                            |
| `level`           | `level`            |             | `AccessLevel.Private \| AccessLevel.Protected \| AccessLevel.Public` | `AccessLevel.Public`                                   |
| `path`            | `path`             |             | `string`                                                             | `undefined`                                            |
| `placeholderHint` | `placeholder-hint` |             | `string`                                                             | `I18n.get(Translations.IMAGE_PICKER_PLACEHOLDER_HINT)` |
| `track`           | `track`            |             | `boolean`                                                            | `undefined`                                            |


## Dependencies

### Depends on

- [amplify-photo-picker](../amplify-photo-picker)

### Graph
```mermaid
graph TD;
  amplify-s3-image-picker --> amplify-photo-picker
  amplify-photo-picker --> amplify-section
  amplify-photo-picker --> amplify-picker
  amplify-photo-picker --> amplify-photo-placeholder
  amplify-photo-picker --> amplify-button
  amplify-picker --> amplify-button
  style amplify-s3-image-picker fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
