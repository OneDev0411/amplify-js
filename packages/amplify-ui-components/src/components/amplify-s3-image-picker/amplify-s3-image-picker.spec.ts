import { newSpecPage } from '@stencil/core/testing';
import { AmplifyS3ImagePicker } from './amplify-s3-image-picker';

describe('amplify-s3-image spec:', () => {
  describe('Component logic ->', () => {
    let amplifyS3ImagePicker;

    beforeEach(() => {
      amplifyS3ImagePicker = new AmplifyS3ImagePicker();
    });

    it('`contentType` should be set to `binary/octet-stream` by default', () => {
      expect(amplifyS3ImagePicker.contentType).toBe('binary/octet-stream');
    });

    it('`level` should be set to `public` by default', () => {
      expect(amplifyS3ImagePicker.level).toBe('public');
    });
  });

  describe('Render logic ->', () => {
    it(`should render no img element without 'imgKey' or 'path'`, async () => {
      const page = await newSpecPage({
        components: [AmplifyS3ImagePicker],
        html: `<amplify-s3-image-picker />`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });
});
