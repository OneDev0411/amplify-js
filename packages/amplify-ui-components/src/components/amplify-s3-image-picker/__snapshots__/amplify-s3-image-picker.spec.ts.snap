// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-s3-image spec: Render logic -> should render no img element without 'imgKey' or 'path' 1`] = `
<amplify-s3-image-picker>
  <!---->
  <slot-fb name="photo-picker">
    <amplify-photo-picker buttontext="Upload" headerhint="Ancilliary text or content may occupy this space here" headertitle="Add Profile Photos" placeholderhint="Placeholder hint"></amplify-photo-picker>
  </slot-fb>
</amplify-s3-image-picker>
`;
